Vue meta/improve props extraction

stafyniaksacha opened the pull request:
Extract nested/union properties from typescript TypeChecker

<details>
<summary><strong>Component input (with complex props)</strong></summary>

```vue
<script setup lang="ts">
export type MyComponentShapes = 'straight' | 'rounded' | 'curved'
export type MyComponentIcon = 'dot' | 'chevron' | 'plus'

export interface MyComponentNestedItem {
  title: string
  content: string
}

class ShapeWithIcon {
  /**
   * The shape of the ShapeWithIcon property.
   */
  public title: string
  public shape: MyComponentShapes
  public actionIcon: MyComponentIcon

  constructor(
    title: string,
    shape: MyComponentShapes,
    actionIcon: MyComponentIcon
  ) {
    this.title = title
    this.shape = shape
    this.actionIcon = actionIcon
  }
}

enum SizeEnum {
  Small,
  Medium,
  Large,
}

const categories = [
  'Uncategorized',
  'Content',
  'Interaction',
  'Display',
  'Forms',
  'Addons',
] as const

type Categories = typeof categories[number]

interface MyComponentProps {
  /**
   * Array
   */
  items: MyComponentNestedItem[]
  /**
   * Interface with union
   */
  primaryItem: MyComponentNestedItem & {
    isPrimary: boolean
    alwaysOpen: boolean
    size: SizeEnum
  }
  /**
   * Class
   */
  action: ShapeWithIcon
  /**
   * Enum
   */
  size: SizeEnum
  /**
   * Context String Literal
   */
  activeCategories: Categories[]
  /**
   * Primtive Type
   */
  count: number
  /**
   * Primtive Type (optional)
   */
  total?: number
}

withDefaults(defineProps<MyComponentProps>(), {})
</script>
```

</details>


<details>
<summary><strong>Output meta</strong></summary>

```json
{
    "name": "MyComponent",
    "global": false,
    "props": [
      {
        "name": "items",
        "documentationComment": "Array",
        "isOptional": false,
        "type": "MyComponentNestedItem[]",
        "schema": {
          "kind": "array",
          "type": "MyComponentNestedItem[]",
          "schema": [
            {
              "kind": "object",
              "type": "MyComponentNestedItem",
              "schema": {
                "title": {
                  "name": "title",
                  "documentationComment": "",
                  "isOptional": false,
                  "type": "string",
                  "schema": "string"
                },
                "content": {
                  "name": "content",
                  "documentationComment": "",
                  "isOptional": false,
                  "type": "string",
                  "schema": "string"
                }
              }
            }
          ]
        }
      },
      {
        "name": "primaryItem",
        "documentationComment": "Interface with union",
        "isOptional": false,
        "type": "MyComponentNestedItem & { isPrimary: boolean; alwaysOpen: boolean; size: SizeEnum; }",
        "schema": {
          "kind": "object",
          "type": "MyComponentNestedItem & { isPrimary: boolean; alwaysOpen: boolean; size: SizeEnum; }",
          "schema": {
            "title": {
              "name": "title",
              "documentationComment": "",
              "isOptional": false,
              "type": "string",
              "schema": "string"
            },
            "content": {
              "name": "content",
              "documentationComment": "",
              "isOptional": false,
              "type": "string",
              "schema": "string"
            },
            "isPrimary": {
              "name": "isPrimary",
              "documentationComment": "",
              "isOptional": false,
              "type": "boolean",
              "schema": {
                "kind": "enum",
                "type": "boolean",
                "schema": [
                  "false",
                  "true"
                ]
              }
            },
            "alwaysOpen": {
              "name": "alwaysOpen",
              "documentationComment": "",
              "isOptional": false,
              "type": "boolean",
              "schema": {
                "kind": "enum",
                "type": "boolean",
                "schema": [
                  "false",
                  "true"
                ]
              }
            },
            "size": {
              "name": "size",
              "documentationComment": "",
              "isOptional": false,
              "type": "SizeEnum",
              "schema": {
                "kind": "enum",
                "type": "SizeEnum",
                "schema": [
                  "SizeEnum.Small",
                  "SizeEnum.Medium",
                  "SizeEnum.Large"
                ]
              }
            }
          }
        }
      },
      {
        "name": "action",
        "documentationComment": "Class",
        "isOptional": false,
        "type": "ShapeWithIcon",
        "schema": {
          "kind": "object",
          "type": "ShapeWithIcon",
          "schema": {
            "title": {
              "name": "title",
              "documentationComment": "The shape of the ShapeWithIcon property.",
              "isOptional": false,
              "type": "string",
              "schema": "string"
            },
            "shape": {
              "name": "shape",
              "documentationComment": "",
              "isOptional": false,
              "type": "MyComponentShapes",
              "schema": {
                "kind": "enum",
                "type": "MyComponentShapes",
                "schema": [
                  "\"straight\"",
                  "\"rounded\"",
                  "\"curved\""
                ]
              }
            },
            "actionIcon": {
              "name": "actionIcon",
              "documentationComment": "",
              "isOptional": false,
              "type": "MyComponentIcon",
              "schema": {
                "kind": "enum",
                "type": "MyComponentIcon",
                "schema": [
                  "\"dot\"",
                  "\"chevron\"",
                  "\"plus\""
                ]
              }
            }
          }
        }
      },
      {
        "name": "size",
        "documentationComment": "Enum",
        "isOptional": false,
        "type": "SizeEnum",
        "schema": {
          "kind": "enum",
          "type": "SizeEnum",
          "schema": [
            "SizeEnum.Small",
            "SizeEnum.Medium",
            "SizeEnum.Large"
          ]
        }
      },
      {
        "name": "activeCategories",
        "documentationComment": "Context String Literal",
        "isOptional": false,
        "type": "(\"Uncategorized\" | \"Content\" | \"Interaction\" | \"Display\" | \"Forms\" | \"Addons\")[]",
        "schema": {
          "kind": "array",
          "type": "(\"Uncategorized\" | \"Content\" | \"Interaction\" | \"Display\" | \"Forms\" | \"Addons\")[]",
          "schema": [
            {
              "kind": "enum",
              "type": "\"Uncategorized\" | \"Content\" | \"Interaction\" | \"Display\" | \"Forms\" | \"Addons\"",
              "schema": [
                "\"Uncategorized\"",
                "\"Content\"",
                "\"Interaction\"",
                "\"Display\"",
                "\"Forms\"",
                "\"Addons\""
              ]
            }
          ]
        }
      },
      {
        "name": "count",
        "documentationComment": "Primtive Type",
        "isOptional": false,
        "type": "number",
        "schema": "number"
      },
      {
        "name": "total",
        "documentationComment": "Primtive Type (optional)",
        "isOptional": true,
        "type": "number | undefined",
        "schema": {
          "kind": "enum",
          "type": "number | undefined",
          "schema": [
            "undefined",
            "number"
          ]
        }
      }
    ],
    "events": [
      {
        "parametersType": "any[]",
        "parameters": [
          {
            "name": "TODO",
            "type": "any",
            "isOptional": "TODO"
          }
        ],
        "documentationComment": ""
      }
    ],
    "slots": []
  },
```

</details>


Todo:
- [ ] extract default values (@johnsoncodehk any idea where we can found them?)
- [ ] extract props/slots (another pr)

elevatebart commented on packages/vue-component-meta/src/index.ts:
```diff
@@ -99,14 +101,55 @@ export function createComponentMetaChecker(tsconfigPath: string) {
+
+					return {
+						name: prop.escapedName,
+						documentationComment: ts.displayPartsToString(prop.getDocumentationComment(typeChecker)),
```
@johnsoncodehk @stafyniaksacha I suggest we use standard naming here to facilitate cross usage.

[react-docgen](https://github.com/reactjs/react-docgen) and [vue-docgen-api](https://vue-styleguidist.github.io/docs/Docgen.html) both use the word description here. 

Since we are changing the API foot print anyway, I suggest we make this align with the rest.

```suggestion
						description: ts.displayPartsToString(prop.getDocumentationComment(typeChecker)),
```

elevatebart commented on packages/vue-component-meta/src/index.ts:
```diff
@@ -99,14 +101,55 @@ export function createComponentMetaChecker(tsconfigPath: string) {
+					return {
+						name: prop.escapedName,
+						documentationComment: ts.displayPartsToString(prop.getDocumentationComment(typeChecker)),
+						isOptional: !!prop.declarations?.[0]?.questionToken,
```
With the same consistency spirit in mind, `required` is more common that `isOptional`

```suggestion
						required: !prop.declarations?.[0]?.questionToken,
```

elevatebart commented:
Hello @stafyniaksacha, 

Would you be so kind as to add a test for your use case?
It is very well explained in the PR description.
Once the PR is merged the only code that remains is what's in the tests.

I just don't want to have your work go to waste.

stafyniaksacha commented:
Hi @elevatebart 

Thanks for your feedback! I made the changes on `description` and `required` fields!
Also added the tests 💪

elevatebart reviewed (approved):
I like it

stafyniaksacha commented:
- Added `events` kind in schema (call signature extraction)
- Added `tags` (tsdoc tag extractions, see [example]( https://github.com/cssninjaStudio/volar/blob/vue-meta/improve-extraction/packages/vue-test-workspace/vue-component-meta/reference-type-props/my-props.ts#L29-L37) and [output](https://github.com/cssninjaStudio/volar/blob/vue-meta/improve-extraction/packages/vue-component-meta/tests/index.spec.ts#L34-L51))

johnsoncodehk commented:
Look great, thank you!
